### ISSUE TITLE
forall/2 bug

### ISSUE THREAD

**Issue by pmoura** [infradig/trealla#258]
Found a bug in the implementation of the `forall/2` predicate, which can be defined in Prolog as:

```logtalk
:- meta_predicate(forall(0, 0)).

forall(Generate, Test) :-
	\+ (Generate, \+ Test).
```

The bug is easy to reproduce:

```text
$ tpl
Trealla Prolog (c) Infradig 2020-2021, v1.11.16-3-gf1753
?- use_module(library(lists)).
true.
?- forall(member(X,[1,1,2]), X == 1).
true.
```

The second query is expected to fail, of course. This bug explains some of the Logtalk distribution test failures, e.g. in the `inlining` example.

**Comment by pmoura** [comment]
Pushed to the Logtalk git repo an additional test for the de facto standard `forall/2` predicate. This test currently fails with Trealla do to the bug described here. The test set is located at `logtalk3/tests/prolog/predicates/forall_2`.

**Comment by infradig** [comment]
The problem is with the (double) negation, for some reason the second time
through it doesn't work properly. If you replace the native (\+)/1 with the
in-Prolog traditional definition it works. I'd happily go with that but
unhappily it's twice as slow (chess.pl goes from 16s to 31s) although it
doesn't affect Logtalk speed.

**Comment by pmoura** [comment]
I see that you fixed this bug in v1.11.18. Thanks! All the tests for the Logtalk `inilining` example now pass. Same for the compliance tests for the `forall/2` predicate. In the case of that example, the bug actually affected the Logtalk compiler itself, resulting in the generation of wrong code.
